src/azure-cli/azure/cli/command_modules/cdn/custom.py

```diff
@@ -71,6 +91,304 @@ def update_endpoint(instance,
+                           custom_fragment, source_pattern, destination, preserve_unmatched_path)
+    if action is not None:
+        actions.append(action)
+
+    rule = DeliveryRule(
+        name=rule_name,
+        order=order,
+        conditions=conditions,
+        actions=actions
+    )
+
+    policy.rules.append(rule)
+    params = EndpointUpdateParameters(
+        delivery_policy=policy
+    )
+
+    return client.endpoints.update(resource_group_name, profile_name, endpoint_name, params)
+
+
+def add_condition(client, resource_group_name, profile_name, endpoint_name,
+                  rule_name, match_variable, operator, match_values=None, selector=None,
+                  negate_condition=None, transform=None):
+
+    endpoint = client.endpoints.get(resource_group_name, profile_name, endpoint_name)
+    policy = endpoint.delivery_policy
+    condition = create_condition(match_variable, operator, match_values, selector, negate_condition, transform)
+    for i in range(0, len(policy.rules)):
+        if policy.rules[i].name == rule_name:
+            policy.rules[i].conditions.append(condition)
+
+    params = EndpointUpdateParameters(
+        delivery_policy=policy
+    )
+
+    return client.endpoints.update(resource_group_name, profile_name, endpoint_name, params)
+
+
+def add_action(client, resource_group_name, profile_name, endpoint_name,
+               rule_name, action_name, cache_behavior=None, cache_duration=None,
+               header_action=None, header_name=None, header_value=None, query_string_behavior=None,
+               query_parameters=None, redirect_type=None, redirect_protocol=None, custom_hostname=None,
+               custom_path=None, custom_querystring=None, custom_fragment=None, source_pattern=None,
+               destination=None, preserve_unmatched_path=None):
+
+    endpoint = client.endpoints.get(resource_group_name, profile_name, endpoint_name)
+    policy = endpoint.delivery_policy
+    action = create_action(action_name, cache_behavior, cache_duration, header_action, header_name,
+                           header_value, query_string_behavior, query_parameters, redirect_type,
+                           redirect_protocol, custom_hostname, custom_path, custom_querystring,
+                           custom_fragment, source_pattern, destination, preserve_unmatched_path)
+    for i in range(0, len(policy.rules)):
+        if policy.rules[i].name == rule_name:
+            policy.rules[i].actions.append(action)
+
+    params = EndpointUpdateParameters(
+        delivery_policy=policy
+    )
+
+    return client.endpoints.update(resource_group_name, profile_name, endpoint_name, params)
+
+
+def remove_rule(client, resource_group_name, profile_name, endpoint_name, rule_name):
+
+    endpoint = client.endpoints.get(resource_group_name, profile_name, endpoint_name)
+    policy = endpoint.delivery_policy
+    if policy is not None:
+        for rule in policy.rules:
+            if rule.name == rule_name:
```

Comment:
if you don't want to throw an exception, better to have a warning message. It's normal for customers to have some typo errors in their command. If you do not give any message, it's quite hard for user to figure out why they just cannot delete anything.

Reply:
Hi @MyronFanQiu 

-  We are planning to support more transforms in the near future so Zunli suggested me to make it nargs='+'.

- What is the standard syntax for warning message? Could you give me an example

Reply:
ok. we can keep `nargs='+'`. Well. we don't have a standard syntax for message. You can use `logger.warnning` to output something like `rule cannot be found. This command will be skipped. Please check the rule name` or something like this.

Reply:
meanwhile, please rebase your branch.

- git checkout dev
- git pull upstream dev
- git checkout hata/rulesEngine
- git rebase dev

Reply:
I will sign off this PR after you rebase it and add warning message. If I can merge it, I will do it. If I cannot, I will ask @Juliehzl to merge this one.

Reply:
@MyronFanQiu  I have added warning messages and resolved the conflict in HISTORY.rst